### PR TITLE
[python] Extend #3354 to categoricals of arbitrary value type

### PR DESCRIPTION
**Pull request by johnkerl**
Follow-up frrom https://github.com/single-cell-data/TileDB-SOMA/pull/3354#discussion_r1876884945 for #3353 [[sc-59407]](https://app.shortcut.com/tiledb-inc/story/59407)

### CODE REVIEW

**Review by bkmartinjr — CHANGES_REQUESTED** [review]
We need to add a couple of unit tests to make sure it handles the most likely cases of non-string categoricals.  I have occasionally seen in real-world data:
* categorical of float/int
* categorical of bool (which obviously should not trigger the conversion :-)

**Comment by johnkerl** [comment]
@bkmartinjr ready for round two!

**Review by bkmartinjr — CHANGES_REQUESTED** [review]
I believe you can save a copy of each column in the cat->no_cat conversion (and likely the converse as well).  left a note inline to that effect.

Also, is it ever possible that this new conversion code will recieve a table with more than one chunk?  If so, it may result in columns with varied chunk counts, which will then fail when you turn it back into a table. While the ingest code path (i.e., from an AnnData.obs Pandas dataframe) will only have a single chunk, it would be good to either:
* handle multiple chunks explicitly
* or, outlaw them and assert that they won't happen ever
Currently if they do happen, it will result in an error when you try to recompose the modified table, as you will have columns with varied chunkiness.

**Comment by johnkerl** [comment]
@bkmartinjr ready for the next round of review!
https://github.com/single-cell-data/TileDB-SOMA/pull/3415/commits/079ed30f182ebf5bd7c6e29051fbb031ddcc50bb

File: apis/python/src/tiledbsoma/io/ingest.py [hunk 1]
```diff
@@ -1170,24 +1158,43 @@ def is_str_cat_col(field: pa.Field) -> bool:
...
+                # for it, but each chunk does have a .dictionary_decode().
+
+                if isinstance(column, pa.ChunkedArray):
+                    column = pa.chunked_array(
+                        [chunk.dictionary_decode() for chunk in column.chunks]
+                    )
+                elif isinstance(column, pa.DictionaryArray):
+                    column = column.dictionary_decode()
+                else:
+                    column = column.to_pylist()
```
Comment (bkmartinjr) [inline apis/python/src/tiledbsoma/io/ingest.py:1175]: I'm still confused why this branch needs to round-trip through a Python list.  As I understand the code, you should be able to just leave it as-is, i.e., as an arrow Array.   `Tables.from_pydict` will happily consume that....

Comment (bkmartinjr) [reply apis/python/src/tiledbsoma/io/ingest.py:1175]: Actually, thinking about this more -- under what conditions will this branch ever get used?   The column is a dict type, and the first two branches cover chunked and not chunked.  Can this ever execute?

If not, it should be an assert/failure, right, as it indicates that something unexpected made it past line 1163

File: apis/python/src/tiledbsoma/io/ingest.py [hunk 2]
```diff
@@ -1170,24 +1158,43 @@ def is_str_cat_col(field: pa.Field) -> bool:
...
-                )
+                if isinstance(column, pa.ChunkedArray):
+                    column = pa.chunked_array(
+                        [chunk.dictionary_encode() for chunk in column.chunks]
+                    )
+                elif isinstance(column, pa.DictionaryArray):
+                    column = column.dictionary_encode()
+                else:
+                    column = pa.array(
+                        column.to_pylist(),
```
Comment (bkmartinjr) [inline apis/python/src/tiledbsoma/io/ingest.py:1190]: see comments above - under what cases does this get used?  (aka why does it need to be round-tripped through Python list?)

File: apis/python/src/tiledbsoma/io/ingest.py [hunk 3]
```diff
@@ -1170,24 +1158,43 @@ def is_str_cat_col(field: pa.Field) -> bool:
...
+                # Convert from categorical to non-categorical.  Note that if
+                # this is a pa.ChunkedArray, there is no .dictionary_decode()
+                # for it, but each chunk does have a .dictionary_decode().
+
+                if isinstance(column, pa.ChunkedArray):
+                    column = pa.chunked_array(
+                        [chunk.dictionary_decode() for chunk in column.chunks]
+                    )
+                elif isinstance(column, pa.DictionaryArray):
+                    column = column.dictionary_decode()
```
Comment (bkmartinjr) [inline apis/python/src/tiledbsoma/io/ingest.py:1173]: coverage indicates missing test for this branch, and a few others

File: apis/python/src/tiledbsoma/io/conversions.py [hunk 4]
```diff
@@ -26,7 +26,7 @@
 _MT = TypeVar("_MT", NPNDArray, sp.spmatrix, PDSeries)
 _str_to_type = {"boolean": bool, "string": str, "bytes": bytes}
 
-STRING_DECAT_THRESHOLD = 4096
+COLUMN_DECAT_THRESHOLD = 4096
```
Comment (bkmartinjr) [inline apis/python/src/tiledbsoma/io/conversions.py:29]: how was this threshold arrived at?  Might be good to document the thinking?

My thoughts:  there are two issues, the size of the encoded value (category enum) and the "overhead" of loading enums in Core.  There is no extra enum "cost" until you exceed max(signed int16). And the enum overhead only gets punitive in very large numbers (e.g. the motivating test case had 4M enums).

TL;DR - I'd be inclined to make this value `max(int16)` (32767) unless you have evidence that 4096 is warranted due to the second concern (enum overhead in Core).

**Review by bkmartinjr — CHANGES_REQUESTED** [review]
sorry!  A couple of more questions/comments in the latest.  Happy to chat IRT if helpful.

File: apis/python/src/tiledbsoma/io/conversions.py (hunk 4 above)
Comment (johnkerl) [reply apis/python/src/tiledbsoma/io/conversions.py:29]: > how was this threshold arrived at? Might be good to document the thinking?

It was your original suggestion. Happy to go with newer suggestions from you as you develop new suggestions, as well as to include the reasoning you provide. Thanks @bkmartinjr !

Comment (bkmartinjr) [reply apis/python/src/tiledbsoma/io/conversions.py:29]: oops -- well, short of some benchmarking, I''m not sure we know the perfect value.  But I doubt that there is much loading overhead (ie.., core overhead) difference between 4096 and 32767, so you might as well start with max(int16) to fully utilize the two bytes.  Apologies for my waffling!

File: apis/python/src/tiledbsoma/io/ingest.py (hunk 3 above)
Comment (johnkerl) [reply apis/python/src/tiledbsoma/io/ingest.py:1173]: @bkmartinjr this is operating on Arrow data as a phase within `tiledbsoma.io.from_anndata` / `tiledbsoma.io.from_h5ad`. I don't know how to tweak AnnData objects such that when they're converted to Arrow, their enum columns are fiddle-forced to be either `pa.DictionaryArray` or `pa.ChunkedArray`.

I'll create a separate whitebox test case to test only this routine, with manually manipulated/force-fiddled Arrow columns.

Comment (bkmartinjr) [reply apis/python/src/tiledbsoma/io/ingest.py:1173]: got it...the only way I can think to do it is to unit test this function directly (for these cases).   (wasn't sure if that was the same as your white box test?)

Comment (johnkerl) [reply apis/python/src/tiledbsoma/io/ingest.py:1173]: @bkmartinjr I'm working on it & hope to have it ready for your review soon -- thanks for waiting!

Comment (johnkerl) [reply apis/python/src/tiledbsoma/io/ingest.py:1173]: And yes, "whitebox test" for me means "directly test internals that aren't user-facing"

**Review by bkmartinjr — APPROVED** [review]
super clean - thanks for enduring the iterations on this!

**Comment by johnkerl** [comment]
Thanks @bkmartinjr !
